통계 api

sayapin1 opened the pull request:
## 이슈 번호
#1 통계
<br>

## 💡 변경 이유

<br>

## 🔑 주요 변경사항
### 1. 통계 api 완료(test code 제외)
- statistics DTO 생성
	- 기존의 create-statistic.dto와 update-statistic.dto는 통계를 생성과 수정하지 않아 제거하고 새로운 dto를 만들었습니다.
- statistics controller 및 service완료
	- 기존에 말씀 드렸던 것과 달리 일자별과 시간별 통계가 출력 값이 다르게 나와야 하는 문제 때문에 서비스 레이어에서 코드를 분리하였습니다. 
	- 또한 쿼리로 들어오는 값이 없을 때 과제에 명시된 대로 기본값이 들어오도록 하였습니다. 다만, jwt에서 사용자 계정을 받아야 해시태그 기본값을 받을 수 있어서 아직 머지하기 전인 현재는 '태그'라는 기본값만 두었습니다. 추후 머지 후 수정하겠습니다.
	- 서비스 코드에서 너무 길어진 부분이 제 스스로도 있다고 느끼지만 실력이 부족해 어떻게 줄여야 될지 모르겠습니다. 여러분들의 의견 기다리겠습니다!
- test code 미작성
	- 빠른 리뷰를 받고 싶어 일단 테스트코드 없이 PR을 열었습니다. 기다리면서 계속 작업하겠습니다.

- 좀 다른 얘긴데 제가 PR 템플릿 건드린 것이 계속 다른 분들께도 영향을 끼친 것 같아... 다시 원상 복구했습니다...
- post repository를 statistics service코드에 바로 주입하지 않고 postLib와 연결.  statisticsService > postLib > postService > postRepository 로 연결되도록 설정.

추가 수정사항
- dto에서 원래 IsOptional 데코레이터를 통해 빈값으로 들어오는 value들에 대해 에러를 던지지 않도록 처리했는데 main과 머지 이후부터 갑자기 IsOptional이 작동을 안하고 있어서.. 현재 dto의 데코레이터들을 대부분 지웠습니다... 이상하게 hashtag만 작동해서 그것만 냅뒀어요....

<br>

## 📷 테스트 결과

sayapin1 committed:
[#1]feat: statistic dto 생성

sayapin1 committed:
[#1]feat: statistic controller 1차 작업

sayapin1 committed:
[#1]feat: statistic service 1차 작업 및 service에서 repository를 사용하기 위해 module에 import함

sayapin1 committed:
[#1] feat: statistics date/hour로 service코드 작업 2차

sayapin1 committed:
[#1] feat: statistics에 쿼리가 없을 때 들어오는 기본값 추가

sayapin1 committed:
[#1] feat: statistics에 쿼리가 없을 때 들어오는 기본값 추가

sayapin1 committed:
[#1] feat: statistics에 쿼리가 없을 때 들어오는 기본값 추가

kangssu commented on src/feature/statistics/dto/statistics.dto.ts:
```diff
@@ -0,0 +1,14 @@
+export class StatisticsDTO {
+  hashtag: string;
+
+  value: 'count' | 'view_count' | 'like_count' | 'share_count';
```
이거 유니온 타입보다 enum으로 빼버리는건 어떤가요?

kangssu replied:
만약 enum으로 빼는게 괜찮다면 아래 2가지 방법 중 괜찮으신 걸로 빼주시면 될 것 같아요!
1. Dto 내부 파일에 enum 작성 : 내부에서만 사용할 것 같다.
```
import { IsNotEmpty } from 'class-validator';

enum StatisticsValueType {
 count ='count',
... 생략
}
export class StatisticsDTO {
... 생략
}
```
2. 별도의 enum 파일 작성 : 지금은 사용하진 않지만 추후에 다른 곳에서 사용하게 될 것 같다.

sayapin1 replied:
다른 곳에서 사용하게 되지는 않을 것 같습니당😂 enum으로 빼는 건 좋은 생각인 것 같아, 1번 방법으로 수정해놓을게요!

kangssu replied:
@sayapin1 
앗하! 넵 😄 괜찮을 것 같아용!

sayapin1 replied:
다른 곳에서 사용하지 않을 거라고 생각했는데 지금 보니 컨트롤러 레이어에 있는 if문 때문에 valueType을 import를 해야되네요ㅋㅋㅋㅋ

kangssu commented on src/feature/statistics/dto/statistics.dto.ts:
```diff
@@ -0,0 +1,14 @@
+import { IsNotEmpty } from 'class-validator';
+
+export class StatisticsDTO {
```
Dto 에서도 데코레이터 전부 적용해주세요! 
또한, 해당 값들이 선택 값인지, 필수 값인지를 표시하는 !:(필수) 또는 ?:(선택) 추가해주세요!
그 이유는 Dto를 봤을때 어떤게 선택값이고 필수값인지를 작업한 사람은 정확히 알겠지만,
협업하는 사람들도 파악해야 하기 때문이예요! 필수는 생략해도 되긴 하지만 저는 명시하는 것이 더 좋다고 생각합니당 😄 

예를들어 아래 형식으로용
```
@IsString()
hashtag!: string; // hashtag의 값을 필수값으로 들어오게 된다. !: 로 필수값으로 표시했기 때문이다.
```

sayapin1 replied:
네! 좋은 의견입니다 그렇게 해놓을게용

kangssu commented on src/feature/statistics/statistics.controller.ts:
```diff
@@ -1,34 +1,37 @@
-  findAll() {
-    return this.statisticsService.findAll();
-  }
+  getStatistics(@Query() query: StatisticsDTO) {
```
1. **getStatistics 함수명**
getStatistics -> 직역하면 가져온다 / 통계  즉, 통계를 전체 가져온다 뜻이 되는 것 같습니다.
다만, 통계를 전체 가져오는 것이 아닌 쿼리에 담긴 값들로 조건을 걸어서 가져오기 때문에
getStatisticsBy... 이런식으로 By로 어떤것으로부터 가져오겠다는 걸 추가하면 좋을 것 같다고 생각합니당

2. **query 변수명**
@Query() query: StatisticsDTO에서 query의 변수명보단 쿼리로 받은 값들의 통합 명칭? 으로 변경하는 것은 어떨까요? 해당 변수에 대한 값은 통계 관련하여 받은 값들을 담아주기 때문에 query로 조회한 값들이 들어가있는 것이 아니라서 적합해보이진 않아보인다는 작은 의견입니당 😮

sayapin1 replied:
1. 컨트롤러에서 분리되는 게 아닌 서비스 단계에서 분리가 되어 컨트롤러에서는 getStatistics, 서비스에서는 ByDate/Hour로 분리해두었습니다.
2. 제가 설정한 변수명의 의미가 모호하다는 생각이 드네요. statsParams 정도로 고민중입니당

kangssu replied:
1. 아하 이해했습니다! 😄 
2. Params 도 괜찮은 것 같아요! 아니면 전 그냥 statisticsDto 또는 statsResponse?  이런것들도 괜찮아보입니당!

sayapin1 replied:
다른 곳에서도 dto로 바로 갖다 쓰려고  statisticsDto라는 이름으로 변경하였습니당!

kangssu commented on src/feature/statistics/statistics.controller.ts:
```diff
@@ -1,34 +1,37 @@
-    return this.statisticsService.findAll();
-  }
+  getStatistics(@Query() query: StatisticsDTO) {
+    if (!query.start) {
```
start, end, hashtag 관련 if 조건문 3개는 Service 로직에 작성하는게 맞아보입니다!
그 이유는 Controller는 처음 호출하는 입구와도 같다고 생각해서 Controller에 진입해서 "값이 없어! 나가!" 또는 "값이 있네! 그럼 넌 Service로 넘어가!" 약간 이런식으로 생각을 하고 있어용..ㅎㅎ

그래서 결국 시간 계산을 하고 값을 변경하는 로직들이기 때문에 Service가 더 적합해보입니당 😄

sayapin1 replied:
안 그래도 이 부분에 대해서 고민을 했었는데, 저도 희수님과 의견이 비슷해서 컨트롤러가 값이 있는지 없는지를 확인해주는 창구? 같다고 생각을 했었거등요.. 그래서 값이 없네? 그럼 이 값을 가지고 서비스로 가! 약간 이런 느낌으로 코드를 짠 건데 역시 서비스에서 처리를 하는 게 맞을까요?

kangssu replied:
@sayapin1 
넵! 서비스가 나을 것 같습니당 😄 
start, end, hashtag가 좀 더 간단했다면 Controller에서 하셔도 괜찮을 것 같지만 로직이 좀 많이 들어가있다고 생각합니다! 

만약에 그냥 hashtag의 디폴트 값처럼 간단했다면.. 아래같은 방법도 가능할 것 같긴 합니다!
넘겨주는 값을 전부 작성해서 디폴트 값을 차라리 같이 넣어줘버리는 거죠!
이러면 아마.. 값이 없으면 디폴트로 들어갔던 것 같긴해용 정확하진 않아용 😅 
```
 if (query.type === 'hour') {
      return this.statisticsService.getStatisticsByHour({
     start: query.start,
     end: query.end,
     hashtag:  query.hashtag| '태그',
});
    }
```

sayapin1 replied:
사실 아예 dto에서 디폴트 값을 주려고도 생각을 했는데...
`hashtag: string = '태그';`
이런 식으로요.
근데 인식을 못하더라고요... 분명 스택오버플로우에도 nest dto 디폴트값 설정이 가능하다고 나와았는데...

sayapin1 replied:
좀 더 알아보고 안되면 서비스로 옮기겠습니당

sayapin1 replied:
서비스 코드에서 별도 함수로 디폴트 값 주는 코드를 분리하였습니다!

kangssu commented on src/feature/statistics/statistics.service.ts:
```diff
@@ -1,26 +1,269 @@
-  }
+  constructor(
+    @InjectRepository(Post) private postRepository: Repository<Post>,
+    @InjectRepository(Hashtag) private hashtagRepository: Repository<Hashtag>,
```
hashtagRepository 를 사용하는 곳이 있나요..? 검색하면 주입해놓은 것밖에 없어서요! 😮 

그리고 postRepository도 바로 주입해서 사용하는 것이 맞지 않다고 생각합니당..!
NestJS에서는 Module로 캡슐화를 한다고 생각합니다. 
즉, Statistics Module이란 캡슐안에 StatisticsController, StatisticsService, StatisticsRepository를 담고 있습니다.
StatisticsService에서 자기 자신의 StatisticsRepository를 사용하는 것은 당연한 것이지만 
HashtagRepository, PostRepository는 불청객 같다는 생각이 듭니다!

그래서 저는 자기 자신이 아닌 다른 레포를 사용해야 한다면 이때 hashtagLib을 주입하여 사용하는 것이 좋다고 생각합니다! 그렇다면 연결은 this.hashtagLib.getHashtag() 이런식으로 hashtagLib파일로 연결시켜서 hashtagLib파일 내에서 hashtagRepository를 연결하는 것이 맞지 않을까? 라는 생각을 해봤습니당 😄

sayapin1 replied:
1. 네 해시태그는 빼는 게 맞겠어요ㅋㅋㅋ 쓰려나 싶어서 넣었는데 정작 쓸 일이 없네요
2. 너무 맞는 말씀인 것 같아요! postLib파일? 을 만들어서 연결해서 해보겠습니당

sayapin1 replied:
혹시 postLib를 통해 repository를 연결하는 예시 코드나 참고할만한 사이트 있으면 알려주실 수 있을까요?

kangssu replied:
@sayapin1 
아래 원티드 과제 링크인데용! 예를들면 statisticsService > postLib > postRepository로 연결했었어요!
https://github.com/kangssu/wanted-pre-onboarding-backend/blob/main/server/src/feature/job_vacancy/jobVacancy.service.ts

그리고 사실 제가 과제때는 단축시킨 것이.. 있어서 위에는 참고만 하시고 흐름은 아래로 부탁드려도 될까요?
원래는 statisticsService > postLib > postService > postRepository가 맞습니다!

1. statisticsService : postService가 아니기 때문에 postLib을 공유 받아서 post 데이터를 연결하고자함
2. postLib : 해당 파일을 자기 자신이 아닌 다른데도 쓰고 있다는 개념으로 협업할때 이해하기 쉽도록 사용함
3. postService : 자기 자신의 Repository를 주입받아서 데이터에 연결함
4. postRepository : 별도 파일이 아닌.. 보라님 하시는 방식으로 postService에서 주입받아서 사용하면 될 것 같아용

postLib은 postModule의 providers에 주입 시켜주시면 됩니다!

sayapin1 replied:
말씀하신대로 변경하였습니당! 감사합니다

kangssu commented on src/feature/statistics/statistics.service.ts:
```diff
@@ -1,26 +1,269 @@
-    return `This action returns a #${id} statistic`;
-  }
+    if (query.type === 'hour') {
+      const startHour = new Date(query.start);
```
이 부분은 getStatisticsByDate랑 getStatisticsByHour 여기에서 동시 사용하는 것으로 확인되는데용
이 부분을 따로 내부 함수나 유틸클래스로 별도 분리해서 호출하면 어떨까 합니다!
3줄이라도 줄일 수 있을 것 같아서요.. 😳

sayapin1 replied:
감사합니당ㅠㅠ

kangssu commented on src/feature/statistics/statistics.service.ts:
```diff
@@ -1,26 +1,269 @@
+
+    if (query.value === 'count') {
+      result = await this.postRepository
+        .createQueryBuilder('post')
```
혹시 아래 코드 if문에 전부 동일한 쿼리로 들어가는게 맞을까요? 그리고 where 조건문만 변경되고 있는게 맞을까요?
```
.createQueryBuilder('post')
        .leftJoinAndSelect('post.hashtags', 'hashtag')
        .select('DATE(post.created_at)', 'date')
        .addSelect('COUNT(post.id)', 'sum')
        .where('hashtag.hashtag = :hashtag', { hashtag: query.hashtag })
```

만약 맞다면 맨 상단에 위 쿼리를 특정 변수에 담고 조건문안에는 뒤에 where 조건문만 변경하면 될 것 같아서요!
예를들어 아래 부분처럼 수정하는 것은 어떨까요?
```
// 모든 조건문에 중복되는 쿼리 맨 상단에 query 변수에 담기
const query = .createQueryBuilder('post')
        .leftJoinAndSelect('post.hashtags', 'hashtag')
        .select('DATE(post.created_at)', 'date')
        .addSelect('COUNT(post.id)', 'sum')
        .where('hashtag.hashtag = :hashtag', { hashtag: query.hashtag });

// 조건문에 따라 andWhere 조건으로 연결(아래 조건은 count value가 일치할 경우에만 연결됨)
if(query.value === 'count'){
.andWhere(
          'DATE(post.created_at) BETWEEN :startDate AND DATE_ADD(:endDate, INTERVAL 1 DAY)',
          {
            startDate: query.start,
            endDate: query.end,
          },
        );
}

// 마지막으로 쿼리 마지막에 중복되는 groupBy, getRawMany를 한번만 작성하여 이어주기
const result = query.groupBy('DATE(post.created_at)').getRawMany();

return result;

```

sayapin1 replied:
쿼리를 특정 변수에 담는 건 생각지 못했던 방법인데! 그럼 코드를 엄청 줄일 수 있을 것 같아요! 바로 해보겠습니당

sayapin1 replied:
생각해보니, 아예 셀렉트절에서 선택하는 값만 조건문에 따라 변수에 담아서 쿼리에 넣으면 될 것 같아용

kangssu commented on src/feature/statistics/statistics.service.ts:
```diff
@@ -1,26 +1,269 @@
+      }
+    }
+
+    let result;
```
let result 부분이 아래 코드를 보니까 계속 result에 담기는 값이 변경되어서 let 변수를 사용하신 것 같아요!
정말 만~~~~약에 아래 조건에 아무것도 걸리는게 없다면? result는 아무런 값도 가지지 않게 되는 것일까요?

제가 위 쿼리 관련 코멘트 달아드린 방법 사용하시면 let result 는 필요 없어보입니다!

사실 개인적인 의견으로는 let 은 안에 값이 계속 변경되서 의도치 않은 값이 들어가버린다면 .. 이런 생각을 많이 해서 Java에서는 많이 사용했지만 Typescript에서는 const를 주로 사용하고 명시된 값만 가져오도록 사용하고 있습니당 😄  (요건 개인차이 인 것 같습니다!)

sayapin1 replied:
감사합니당 위 방법 참고해서 고쳐보겠습니다

kangssu reviewed (commented):
통계 관련 쿼리 코드보고 많이 배우고 갑니당! 쿼리 지옥이였을 것 같은데 대단하군용 😱 👍

dawwson commented on src/feature/statistics/dto/statistics.dto.ts:
```diff
@@ -0,0 +1,14 @@
+  @IsNotEmpty()
+  type: 'date' | 'hour';
+
+  start: string;
+
+  end: string;
```
날짜 형식인지도 검증이 필요할 것 같습니다. 요구사항에는 yyyy-mm-dd 형식으로 받아야 됨과 동시에 필수값은 아니기 때문에
```
@IsOptional()
@IsDateString() // 또는 @IsISO8601()
start?: string;
```
이렇게 조합해서 쓸 수 있을 거 같습니다.

sayapin1 replied:
네! 데코레이터 적용시켜놓을게용

dawwson commented on src/feature/statistics/statistics.service.ts:
```diff
@@ -1,26 +1,269 @@
+    const startDate = new Date(query.start);
+    const endDate = new Date(query.end);
+    const thirtyDaysInMilliseconds = 30 * 24 * 60 * 60 * 1000;
+    if (endDate.getTime() - startDate.getTime() > thirtyDaysInMilliseconds) {
+      throw new UnprocessableEntityException(
+        '최대 조회 가능한 날짜는 30일입니다.',
+      );
+    }
```
end-start 가 30일 이하여야 한다는 조건이네요..
저는 개인적으로 422 상태코드가 생소한데, dto에서 validation pipe를 거쳐서 400에러를 보낼 수 있도록 하는 것도 한가지 방법이 될 수 있을 거 같아요.
```
@ValidateIf((object, value) => object.end - value < 30일)
start?: string;

@ValidateIf((object, value) => value - object.start < 30일)
end?: string;
```
손코딩이라 정확한 코드는 아니지만,,,ㅎㅎ
class-validator에서 제공하는 데코레이터를 활용해도 좋을 거 같습니다.

이거는 여유 되시면 한 번 고려해봐주세용!!
https://github.com/typestack/class-validator#conditional-validation

sayapin1 replied:
422 는 서버가 요청을 이해했지만 처리할 수 없는 경우 사용된다고 하네요! 그래서 클라이언트의 요청 자체가 잘못된 것은 아니지만 처리할 수 없는 요청, 즉 조회 가능한 시간을 넘어섰을 때에 쓰기에 적절하다 판단되어 사용하였습니다
ValidateIf는 사용해본 적이 없는 데코레이터인데 서비스까지 가기 전에 dto에서 검증을 할 수 있다면 너무 좋은 방법인 것 같아요! 확인해보겠습니다ㅎㅎ

sayapin1 replied:
한번 시도해보려고 했는데, 생각해보니 시간에 대한 디폴트 값을  줘야하는 문제 때문에 좀 더 고민해봐야할 것 같아용..

sayapin1 committed:
[#1] fix: statistics.controller.ts 코드 수정

1. query라는 변수명을 statsParams로 수정
2. dto에서 value의 타입을 enum으로 주었기 때문에 controller에서도 valueType이라는 enum을 import해 사용하도록 수정

sayapin1 committed:
[#1] fix: statistics.dto.ts 코드 수정

1. 데코레이터 적용
2. 해당필드가 선택/필수값인지 표시
3. value의 type을 enum으로 바꾸고 enum을 외부 파일로 분리

sayapin1 committed:
[#1] fix: statistics.module.ts 코드 수정

1. 사용하지 않는 hashtag repository 제거

sayapin1 committed:
[#1] fix: statistics.service.ts 코드 수정

1. 사용하지 않는 hashtagRespository 제거
2. 변수명 query를 statsParams로 변경
3. 쿼리의 셀렉트 절의 선택 컬럼명만 따로 변수명으로 빼고 중복되는 코드는 합침

sayapin1 committed:
[#1] fix: statisticsValueType.enum.ts 추가

1. dto의 value의 type을 enum으로 변경하며 외부 파일로 분리함

laetipark reviewed (approved):
고생 많으십니다! 코드 잘 확인하고 갑니다 :)

kangssu commented on src/feature/statistics/dto/statistics.dto.ts:
```diff
@@ -0,0 +1,30 @@
+export class StatisticsDTO {
+  @IsOptional()
+  @IsString()
+  hashtag: string = '태그';
```
@IsOptional이면 선택값 같은데 ? 요거 빠졌어용!

sayapin1 replied:
어이쿠 제가 어제 기본값 넣어본다고 이것저것 해보다가 지웠나봐용

kangssu commented on src/feature/statistics/statistics.module.ts:
```diff
@@ -1,8 +1,11 @@
+import { Post } from '../../entity/post.entity';
 
 @Module({
+  imports: [TypeOrmModule.forFeature([Post])],
```
여기서는 PostModule을 임포트 해줘야 합니다! 
PostRepository가 StatisticsModule의 자기 자신 Repository가 아니기 때문입니당 😄

sayapin1 replied:
넵넵 postLib로 분리한 후 수정하였습니당

kangssu reviewed (commented):
혹시 어제 PR이 전부 수정된걸까용??

sayapin1 committed:
[#1] fix: statistics.service.ts에서 post레포지토리를 가져오는 것을 postLib를 통해 연결함.

1. statisticsService > postLib > postService > postRepository 흐름으로 연결

sayapin1 committed:
[#1] fix: statistics.controller.ts의 if조건문을 서비스 코드로 옮김

1. 서비스 코드로 옮기면서 각 함수 안에 넣기에는 중복되는 코드가 많아 별도의 함수로 분리함.

sayapin1 committed:
Merge branch 'main' into feature/posts/statistics/1-statistics

sayapin1 committed:
[#1] fix: main -> statistics로 머지 후 충돌 파일들 수정 및 dto수정

1. 미리 충돌을 해결하기 위해 main에서 statistics로 파일들 머지 이후 충돌 파일들 및 중복으로 복사된 코드들 수정
2. dto의 isOptional 데코레이터가 있음에도 빈값으로 보낼 때 유효성 검사에 걸리게 되어 일단 전부 제거(hashtag 제외)

sayapin1 committed:
[#1] fix: 엔티티 주석달아놓은 거 제거함..

kangssu reviewed (approved):
늦은시간까지 😢 너무 고생하셨습니다!
